### PR TITLE
Add more branch paths for logins migrations

### PR DESCRIPTION
**Pull request by skhamis**
Mainly building off of the previous implementation of `MigrationPlan`. Added the rest of the branches that migrations could go with accompanying tests. 

Things now go through three phases:
1. Loop through both tables and fill `MigrationPlan` accordingly
2. Loop through `MigrationPlan` and apply any fixups and/or determine if we need to fix any "out of sync" issue (i.e if we fix up a mirror without a local, we move it to the local table)
3. Insert logins into their respective DBs and track proper metrics

It might increase how many loops but we only need to read from the DB once so I think the perf hit is minimized and gives us the best chance to reduce data loss

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (mhammond) [inline components/logins/src/migrate_sqlcipher_db.rs:254]: This pattern repeats a bit, but it's not clear how to avoid it - but maybe see what you can come up with - anything we can do to reduce the lines in this large function would help (although there is a balance to strike - eg, splitting it up into discrete functions would likely make everything more complex rather than less)

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (mhammond) [inline components/logins/src/migrate_sqlcipher_db.rs:420]: This should be `if let` (I expect clippy will complain about that too, and will probably tell you exactly what to replace it with too :) I suspect that's what you were getting at in the comment, but I think it will work fine (and if so, the comment should also be removed)

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (mhammond) [inline components/logins/src/migrate_sqlcipher_db.rs:424]: That enum variant is declared as `Skip(String)`, so I think `err` is already a string?

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (mhammond) [inline components/logins/src/migrate_sqlcipher_db.rs:449]: I would have though `if let Err(err) =...` would work here too (any match with 2 arms and where 1 is empty is suspect)

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (mhammond) [inline components/logins/src/migrate_sqlcipher_db.rs:508]: a bit of a nit, but I don't think `migrate` is great here as part of the name - maybe replace `migrate` with `insert` here?

**Review by mhammond — APPROVED** [review]
This is looking good to me, thanks! Please make the changes I suggested and wait for at least Lougenia or Ben to approve it too before landing it (and also make whatever changes clippy suggests)

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (bendk) [reply components/logins/src/migrate_sqlcipher_db.rs:254]: For this one and the other repitition, what about switching to `migration_plan.logins.iter_mut()`?  That way I think  you could just change the `MigrationLogin` data directly.  I think it also makes it more clear which data is changing and which is staying the same.

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (bendk) [inline components/logins/src/migrate_sqlcipher_db.rs:291]: Could you combine this and the last method call into `maybe_fixup()`?

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (bendk) [inline components/logins/src/migrate_sqlcipher_db.rs:448]: This is a weird error case.  We had valid data but failed to insert it.  It seems like something really wrong is happening, like the disk is full or failing.  I don't have any suggestions on how to change the code, but I'll just note that it seems likely that inserts are going to fail in general.

**Review by bendk — APPROVED** [review]
I like how you split up the fixup/validation code and the insert code.  I agree with taking care of the things  @mhammond flagged, but I couldn't see anything else important so I'm marking approve.

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (skhamis) [reply components/logins/src/migrate_sqlcipher_db.rs:254]: I'll give `iter_mut` a shot, I was wrestling with the borrow checker quite a few times to try modify the data before I gave up and went for the overriding way.

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (skhamis) [reply components/logins/src/migrate_sqlcipher_db.rs:230]: Thanks for this! I really like the matching as it feels easier to read (even with an additional indentation) and more rust-like -- will definitely go this route

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (skhamis) [reply components/logins/src/migrate_sqlcipher_db.rs:426]: I agree, this is really shouldn't happen in the wild without some much bigger issue present but felt like making sure we can at least attempt to prevent data loss (however unlikely) might save us in some small way.

**Comment by skhamis** [comment]
Thanks for the review all, I tried various ways to reduce the LOC but it either increased complexity or decreased readability.  Right now it's easy-to-follow, easier to remove and I believe it's in a pretty good state right now and we can certainly revisit some of the steps in each of the phases as the dust settles with the entire body of work. Gonna go ahead and merge!

**Pull request merged**
